### PR TITLE
ENG-9796 Add pipeline build command

### PR DESCRIPTION
**Pull request by willdavies0**
# Objective
Add a local build flow for pipeline iteration.

## Changes
- Merge the branch with the latest `main` command model changes
- Align `pipeline build` with the shared selector flow so it can resolve pipelines by alias, pipeline ID, or path while keeping `--path` required
- Create missing draft pipelines, update existing Orchestra-backed drafts, and start the returned draft version through the shared run path
- Add docs and tests for selector-based build behavior, including explicit coverage that alias is optional and generated from `--path` when needed
- Move the shared git warning confirmation prompt into `utils/git.py` and rename it for clearer reuse across commands
- Guard the build create-failure path with an explicit exit after the shared API error renderer
- Reuse the shared upsert success-response parsing helpers instead of duplicating response/body/pipeline-id parsing in `build_pipeline.py`

## Testing
Unit tested and linted

## Checklist
- [x] Verified these changes are backwards compatible (db migrations, model updates)

## Additional Notes

<div><a href="https://cursor.com/agents/bc-a4956dd9-4fc4-462e-b779-06c49b93fe03"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-web-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-web-light.png">Open in Web</picture></a>&nbsp;<a href="https://cursor.com/background-agent?bcId=bc-a4956dd9-4fc4-462e-b779-06c49b93fe03"><picture><source media="(prefers-color-scheme: dark)" srcset="https://cursor.com/assets/images/open-in-cursor-dark.png"><source media="(prefers-color-scheme: light)" srcset="https://cursor.com/assets/images/open-in-cursor-light.png">Open in Cursor</picture></a>&nbsp;</div>

### CODE REVIEW

File: orchestra_cli/src/build_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+import httpx
+import typer
+
+from ..utils.api import auth_headers, fail_with_response, request_or_exit, require_api_key
+from ..utils.constants import get_create_pipeline_url, get_update_pipeline_url
+from ..utils.styling import green, indent_message, red, yellow
+from ..utils.yaml_loader import load_validated_pipeline_data
+from .pipeline_upsert import build_upsert_payload
+from .run_pipeline import _confirm_warnings_or_exit, build_run_payload, start_pipeline_run
```
Comment (willdavies0) [inline orchestra_cli/src/build_pipeline.py]: rename _confirm_warnings_or_exit to confirm_warnings_or_exit

File: README.md [hunk 2]
```diff
@@ -283,6 +315,9 @@ orchestra pipeline run -a finance-etl
 
 # Start a run and exit immediately
 orchestra pipeline run -a finance-etl --no-wait
+
+# Build a draft pipeline from local YAML and start it
+orchestra pipeline build -a finance-etl -p ./pipelines/etl.yaml
```
Comment (willdavies0) [inline README.md]: why is alias and path needed? should be one or the other?
is there some API limitation preventing this

File: orchestra_cli/src/build_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+def _extract_pipeline_version(response: httpx.Response, action: str) -> int:
+    try:
+        body = response.json()
+    except Exception:
+        typer.echo(red(f"❌ {action} failed: success response was not valid JSON"))
+        typer.echo(yellow(indent_message(response.text)))
+        raise typer.Exit(code=1)
+
+    version_number = body.get("latestVersionNumber") or body.get("currentVersionNumber")
```
[the remark this answers is not in the document]
Comment (willdavies0) [reply orchestra_cli/src/build_pipeline.py]: **Written by AI (fully autonomous):** Fixed by preferring latestVersionNumber only when it is present, then falling back to currentVersionNumber on an explicit None check. That preserves valid 0 values instead of treating them as missing. I also added a regression test for the zero-version case.

File: orchestra_cli/src/build_pipeline.py (hunk 1 above)
Comment (willdavies0) [reply orchestra_cli/src/build_pipeline.py]: **Written by AI (fully autonomous):** Renamed the shared warning helper to confirm_warnings_or_exit so it reads as an intentionally shared function rather than command-private implementation detail. Updated both run and build to use the new name.

File: README.md (hunk 2 above)
Comment (willdavies0) [reply README.md]: **Written by AI (fully autonomous):** Clarified the docs here. The local path is still needed because build validates and uploads the local YAML content, while alias is needed because the current API endpoints update and start the remote draft pipeline by alias. If the API later supports an addressable local-only build target, we could revisit that shape.

File: orchestra_cli/src/run_pipeline.py [hunk 4]
```diff
@@ -21,7 +18,7 @@
 from ..utils.styling import bold, green, indent_message, red, yellow
 
 
-def _confirm_warnings_or_exit(force: bool, path: Path | None = None) -> None:
+def confirm_warnings_or_exit(force: bool, path: Path | None = None) -> None:
```
Comment (willdavies0) [inline orchestra_cli/src/run_pipeline.py]: rename this for clarity over what it does and move to utils

File: README.md (hunk 2 above)
Comment (willdavies0) [reply README.md]: hmm yeah but the alias should be optional. we are going to build the alias from the path using `_resolve_generated_alias_selector` this is called resolve_pipeline_selector function. Ensure there are unit tests covering this and update the docs ensuring alias is clearly optional

File: orchestra_cli/src/build_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        return PipelineSelector(alias=str(alias))
+
+    typer.echo(
+        red("❌ Build failed: existing pipeline metadata did not include alias or pipeline id"),
+    )
+    raise typer.Exit(code=1)
+
+
+def _build_create_selector(
+    *,
```
Comment (ojc-orchestra) [inline orchestra_cli/src/build_pipeline.py]: don't need

File: orchestra_cli/src/build_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+def _storage_provider(existing_pipeline: dict[str, object]) -> str | None:
+    for key in ("storage_provider", "storageProvider"):
+        value = existing_pipeline.get(key)
+        if isinstance(value, str):
+            return value
+    return None
+
+
+def _upsert_draft_pipeline(
+    *,
```
Comment (ojc-orchestra) [inline orchestra_cli/src/build_pipeline.py]: don't need

File: orchestra_cli/src/pipeline_upsert.py [hunk 7]
```diff
@@ -30,18 +30,32 @@ def build_upsert_payload(
...
+    return body
+
+
+def require_pipeline_id_from_success_body(
+    body: dict[str, object],
+    action: str,
+) -> str:
     pipeline_id = body.get("id")
+    if pipeline_id is None:
+        pipeline_id = body.get("pipeline_id")
```
Comment (ojc-orchestra) [inline orchestra_cli/src/pipeline_upsert.py]:
```suggestion
pipeline_id = body.get("id") or body.get("pipeline_id")
```

File: orchestra_cli/src/pipeline_upsert.py [hunk 8]
```diff
@@ -50,6 +64,14 @@ def require_pipeline_id_from_success_response(
     return str(pipeline_id)
 
 
+def require_pipeline_id_from_success_response(
+    response: httpx.Response,
+    action: str,
+) -> str:
+    body = require_pipeline_body_from_success_response(response, action)
+    return require_pipeline_id_from_success_body(body, action)
```
Comment (ojc-orchestra) [inline orchestra_cli/src/pipeline_upsert.py]: you can inline body here

File: orchestra_cli/src/run_pipeline.py [hunk 9]
```diff
@@ -117,38 +89,32 @@ def _poll_until_terminal(
...
-    api_key = require_api_key()
-    selector = resolve_pipeline_selector(alias, pipeline_id, path)
-    selector_name = selector.display()
-
-    _confirm_warnings_or_exit(force, path)
-
-    payload = selector.to_payload()
+def build_run_payload(
+    selector: PipelineSelector,
+    *,
```
Comment (ojc-orchestra) [inline orchestra_cli/src/run_pipeline.py]: is this needed?

File: orchestra_cli/src/run_pipeline.py [hunk 10]
```diff
@@ -117,38 +89,32 @@ def _poll_until_terminal(
...
         payload["branch"] = branch
     if commit:
         payload["commit"] = commit
+    if version_number is not None:
+        payload["versionNumber"] = version_number
+    return payload
+
 
+def start_pipeline_run(
+    *,
```
Comment (ojc-orchestra) [inline orchestra_cli/src/run_pipeline.py]: needed?

File: orchestra_cli/src/build_pipeline.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+                    "✅ Draft pipeline "
+                    f"({create_selector.display()}) created as version {version_number}",
+                ),
+            )
+            return PipelineSelector(pipeline_id=pipeline_id), version_number
+
+        fail_with_response("Build", create_response)
+        raise typer.Exit(code=1)
+
+    assert existing_pipeline is not None
```
Comment (ojc-orchestra) [inline orchestra_cli/src/build_pipeline.py]: prefer not to have asserts in prod code

**Comment by willdavies0** [comment]
**Written by AI (assisted):** Addressed review feedback: removed unnecessary `*` keyword-only markers in build/run helpers, collapsed pipeline id extraction in `pipeline_upsert`, inlined the success-body helper, dropped redundant `alias` threading now that `lookup_selector` carries resolution, replaced the post-create assert with an `else` branch, fixed empty-payload handling in `start_pipeline_run`, softened the missing-path error message, expanded README on optional `--alias`, and added a selector unit test for the build-style path. `fail_with_response` already raises `typer.Exit`, so no extra exit was required after lookup failures. All `make test` checks pass.

File: orchestra_cli/src/build_pipeline.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+        get_pipeline_url(),
+        params=selector.to_payload(),
+        timeout=30,
+        headers=auth_headers(api_key),
+    )
+
+    if response.status_code == 404:
+        return None
+    if response.status_code != 200:
+        fail_with_response("Build", response)
```
Comment (ojc-orchestra) [inline orchestra_cli/src/build_pipeline.py]: @willdavies0 is this meant to `raise fail_with_response`?

Comment (willdavies0) [reply orchestra_cli/src/build_pipeline.py]: good spot

**Review by ojc-orchestra — APPROVED** [review]
